### PR TITLE
Get List of Company, Deal Properties

### PR DESCRIPTION
**Pull request by LeeStevens318**
Looks like this may be generic for deals etc too. not sure how you want to structure it. i,e, create common classes and inherit for each type (company, deal etc)

### CODE REVIEW

**Comment by nover** [comment]
Hi @LeeStevens318 

Thanks for submitting this. I'll go have a look at HubSpots documentation in relation to how we should structure this.

But if it holds true that each "type of thing" has it's own endpoints and data-models in HubSpot for managing Properties, I think it is a good approach to bake that into the existing clients we have, rather than creating a `PropertiesClient`.

**Comment by LeeStevens318** [comment]
Yep cool let me know. 

I also need to use the Deal Properties for the project I'm working on now, so happy to implement that too, just thought I'd check with you on structure first :).

**Comment by nover** [comment]
cool, perhaps implementing / working on the deals property integration is going to shed some light no the proper architecture for this? :)

**Comment by LeeStevens318** [comment]
I guess so :). I'll have a go tomorrow, it's late in Oz now.

**Comment by LeeStevens318** [comment]
The models between `[Company](https://legacydocs.hubspot.com/docs/methods/companies/get_company_properties)` and `[Deal](https://legacydocs.hubspot.com/docs/methods/deals/get_deal_properties)` Properties differ slightly.

I created a "Common" folder and then inherited for each type (Didn't want to use the Properties folder because app files are in there, more than happy for you to move around to what you think is best) in their own client folders.

**Comment by LeeStevens318** [comment]
Added nullable check on `Convert.ChangeType` in the `RequestDataConverter` class and method `ConvertSingleEntity` to allow nullable properties i.e. `OwnerId` in `Company` and `Deal`.

Otherwise it would throw exception `Invalid cast from 'System.String' to 'System.Nullable`1[[System.Int64, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]'`

File: src/Common/Dto/Properties/PropertyHubSpotEntity.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        public string FieldType { get; set; }
+
+        [DataMember(Name = "options")]
+        public List<PropertyOptionHubSpotEntity> Options { get; set; }
+
+        public bool IsNameValue => false;
+
+        public string RouteBasePath => "/properties/v1";
+
+        public void ToHubSpotDataEntity(ref dynamic dataEntity)
```
Comment (nover) [inline src/Common/Dto/Properties/PropertyHubSpotEntity.cs:35]: **issue** should be made `virtual` to allow overwriting it in inheriting classes

File: src/Common/Dto/Properties/PropertyHubSpotEntity.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            
+        }
+
+        public void FromHubSpotDataEntity(dynamic hubspotData)
```
Comment (nover) [inline src/Common/Dto/Properties/PropertyHubSpotEntity.cs:40]: **issue** should be made `virtual` to allow overwriting it in inheriting classes

File: src/Common/Dto/Properties/PropertyListHubSpotEntity.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        public bool Remove(T item)
+            => Properties.Remove(item);
+
+        public int Count
+            => Properties.Count;
+
+        public bool IsReadOnly
+            => false;
+
+        public void ToHubSpotDataEntity(ref dynamic dataEntity)
```
Comment (nover) [inline src/Common/Dto/Properties/PropertyListHubSpotEntity.cs:43]: **issue** should be made `virtual` to allow overwriting it in inheriting classes

File: src/Common/Dto/Properties/PropertyListHubSpotEntity.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        {
+
+        }
+
+        public void FromHubSpotDataEntity(dynamic hubspotData)
```
Comment (nover) [inline src/Common/Dto/Properties/PropertyListHubSpotEntity.cs:48]: **issue** should be made `virtual` to allow overwriting it in inheriting classes

File: src/Core/Requests/RequestDataConverter.cs [hunk 5]
```diff
@@ -289,8 +289,11 @@ internal object ConvertSingleEntity(ExpandoObject dynamicObject, object dto)
                 _logger.LogDebug("Have target prop? '{0}' with name: '{1}' and actual value: '{2}'", targetProp != null,
                     dynamicProp.Key, dynamicValue);
 
-                targetProp?.SetValue(dto, dynamicValue.GetType() == targetProp.PropertyType ? dynamicValue : Convert.ChangeType(dynamicValue, targetProp.PropertyType));
+                targetProp?.SetValue(dto, dynamicValue.GetType() == targetProp.PropertyType 
+                    ? dynamicValue 
+                    : Convert.ChangeType(dynamicValue, Nullable.GetUnderlyingType(targetProp.PropertyType) ?? targetProp.PropertyType));
```
Comment (nover) [inline src/Core/Requests/RequestDataConverter.cs:294]: **praise** nice catch 👍

File: src/Deal/Dto/DealHubSpotEntity.cs [hunk 6]
```diff
@@ -27,7 +27,7 @@ public DealHubSpotEntity()
         [DataMember(Name = "pipeline")]
         public string Pipeline { get; set; }
         [DataMember(Name = "hubspot_owner_id")]
-        public string OwnerId { get; set; }
+        public long? OwnerId { get; set; }
```
Comment (nover) [inline src/Deal/Dto/DealHubSpotEntity.cs:30]: **comment** it's fine that we break it, but we need to list it as a breaking change when releasing.

File: src/Deal/Interfaces/IDealHubSpotEntity.cs [hunk 7]
```diff
@@ -8,7 +8,7 @@ public interface IDealHubSpotEntity : IHubSpotEntity
         string Name { get; set; }
         string Stage { get; set; }
         string Pipeline { get; set; }
-        string OwnerId { get; set; }
+        long? OwnerId { get; set; }
```
Comment (nover) [inline src/Deal/Interfaces/IDealHubSpotEntity.cs:11]: **question**

Are we sure that we can _always_ materialize the ownerIds as longs ?

**Review by nover — CHANGES_REQUESTED** [review]
overall I think it looks good but I have a few comments / concerns we should address before merging

File: src/Deal/Interfaces/IDealHubSpotEntity.cs (hunk 7 above)
Comment (nover) [reply src/Deal/Interfaces/IDealHubSpotEntity.cs:11]: I quess we can, just looked at their documentation.

File: src/Common/Dto/Properties/PropertyHubSpotEntity.cs (hunk 1 above)
Comment (LeeStevens318) [reply src/Common/Dto/Properties/PropertyHubSpotEntity.cs]: Updated

File: src/Common/Dto/Properties/PropertyHubSpotEntity.cs (hunk 2 above)
Comment (LeeStevens318) [reply src/Common/Dto/Properties/PropertyHubSpotEntity.cs]: Updated

File: src/Common/Dto/Properties/PropertyListHubSpotEntity.cs (hunk 3 above)
Comment (LeeStevens318) [reply src/Common/Dto/Properties/PropertyListHubSpotEntity.cs]: Updated

File: src/Common/Dto/Properties/PropertyListHubSpotEntity.cs (hunk 4 above)
Comment (LeeStevens318) [reply src/Common/Dto/Properties/PropertyListHubSpotEntity.cs]: Updated

**Review by LeeStevens318 — COMMENTED** [review]
Updated

**Review by nover — APPROVED** [review]

**Pull request merged**